### ISSUE TITLE
[Discussion] Tightening SGLang's RL inner loop API: scheduler drain, IPC weight transfer, sleep/wake reliability

### ISSUE THREAD

**Issue by bzantium** [sgl-project/sglang#24119]
## Background

SGLang has become a common choice as the inference engine inside RL training frameworks (verl, OpenRLHF, NeMo Aligner, slime, and others). The 2026 Q1 roadmap (#12780) explicitly lists "training inference mismatch mitigation" and "SGLang Gateway as the DP scheduler for rollout in the RL framework" as goals, and recent merges such as #6311 have already moved several RL shaped APIs into the engine. This issue collects three closely related gaps that we keep hitting when integrating SGLang into RL inner loops, and sketches a framework agnostic shape so that the same primitives serve verl, slime, and the rest equally well.

The shared theme is the same. RL training repeatedly leaves SGLang's warm serving assumption: the engine is initialized once, weights are loaded once, and requests stream in at a steady rate. RL instead alternates between long generation bursts, an explicit barrier where in flight requests must be flushed, a fast weight transfer from the trainer process, and a resume. Each of the three items below is a place where SGLang either lacks a primitive that this lifecycle needs, or has one whose corner case behaviour keeps leaking back into framework code as patches.

## 1. Scheduler drain / quiesce API

Today there is no public way to ask SGLang "stop accepting new work, finish what you have, and tell me when you are idle." `release_memory_occupation` has the side effect of doing something in this neighbourhood, but it also tears down the working set, which is too heavy for the common case of "I am about to push new weights and want to make sure nothing is mid decode." Frameworks end up either polling internal scheduler state through monkey patches, or carrying a patch into `scheduler_metrics_mixin` to expose a queue length signal.

A clean primitive would look something like

    await engine.drain()       # blocks until in flight requests complete
    engine.update_weights_from_ipc(...)
    engine.resume()

with `drain` returning quickly when there is nothing in flight, and an explicit `paused` flag so that a request submitted during the drain window is rejected (or queued, configurable) rather than silently racing the weight update. This is the smallest piece of the three but it removes a surprising amount of glue code from every RL framework that touches SGLang.

## 2. `update_weights_from_ipc` for the colocated layout

`Engine.update_weights_from_ipc` already exists in main (`python/sglang/srt/entrypoints/engine.py`, around line 980), with the signature

    def update_weights_from_ipc(self, zmq_handles: Dict[str, str], flush_cache: bool = True)

backed by `tokenizer_manager.update_weights_from_ipc` and the checkpoint-engine integration. `update_weights_from_distributed` (NCCL based) covers the cross-node case well, and the recent extension to multiple weight groups (#6311 and v0.4.10) is appreciated. The case that does not yet have a published shape is the same-node colocated layout, where the trainer process and the SGLang worker share GPUs. NCCL works there too, but going through NCCL when both endpoints sit on the same device is wasted bandwidth and serialization cost compared to a `cudaIpcMemHandle` handoff.

The question we would like maintainer input on is whether the existing IPC entrypoint should also accept a raw `cudaIpcMemHandle` dictionary for the colocated case, alongside the ZMQ-handle path it already takes, or whether the ZMQ broker is the intended path even there. Concretely:

    engine.update_weights_from_ipc({
        "model.layers.0.self_attn.q_proj.weight": handle,
        ...
    })

where `handle` is whatever opaque payload SGLang prefers (a `cudaIpcMemHandle_t`, a tuple of handle and shape, or a Python wrapper around `torch.cuda.IPCHandle`), keyed by the same naming the existing `update_weights_from_*` family already uses, would let verl, slime, NeMo Aligner, and OpenRLHF retire their colocated weight transfer patches and converge on one path, and would slot cleanly under the drain primitive proposed above. If the ZMQ broker is the intended path for the colocated case too, this section collapses to documentation pinning that down.

## 3. Sleep / wake reliability

`release_memory_occupation` and `resume_memory_occupation` are the right primitives, and the corner cases reported in #7939 and #6367 turn out to be partly addressed already. PR #10873 added `enable_weights_cpu_backup` on `ServerArgs`, which wraps the weight region in `torch_memory_saver`'s `enable_cpu_backup=True` so that resume restores the model from a CPU mirror; with that flag on, a `release → resume → generate` sequence preserves the model bytes. An earlier framing of this section speculated about a stale CUDA-Graph capture interaction; without a reproducer that isolates capture staleness from the missing-weights effect, that paragraph is better treated as not a concrete ask. What remains of the byte-equivalence piece is a UX question: should `enable_weights_cpu_backup` flip on by default when `enable_memory_saver=True`, or should the engine warn when a user resumes without having either called `update_weights_*` between release and resume or opted into the cpu-backup flag.

The remaining concrete piece is the in-flight / state-machine guard. `Scheduler.release_memory_occupation` already opens with `assert self.is_fully_idle()`, which detects the bad case but crashes the scheduler process with `AssertionError` (and is stripped under `python -O`) instead of returning an error to the caller. `resume_memory_occupation` has no equivalent check. The narrower ask is to lift `success` / `message` onto `ReleaseMemoryOccupationReqOutput` and `ResumeMemoryOccupationReqOutput` the way `UpdateWeightsFromIPCReqOutput` already carries them, replace the assert with a propagated error response, add the symmetric check to resume, and plumb the result through `tokenizer_manager` and the HTTP handlers so callers see a clean failure rather than a scheduler crash. PR #24123 implements this piece; it does not depend on section 1's drain semantics or on the IPC question above. Direct sleep / wake calls during in-flight work should ultimately go through the drain primitive proposed in section 1; until that exists, the propagated error at least makes the failure mode legible.

## Why these three together

They are listed in one issue because they answer the same question from three angles: how does an external framework hand control of an SGLang engine off to a training step and back, without leaking implementation details into the framework. Splitting them risks fixing one piece in isolation while leaving the rest off contract, which is what produced the current per framework patches.

Happy to help with the implementation; opening this as a discussion first because the API shape is the part that benefits most from feedback before code lands. The first concrete piece (the section 3 in-flight / state-machine guard) is up at #24123, since it is independent of the design questions in sections 1 and 2. We are willing to follow up with the drain primitive (section 1) and the IPC question (section 2) as separate PRs in whichever order the discussion settles on.

**Comment by SohamRajpure** [comment]
May I help with this? Currently working on Ray side to integrate RL control plane and weight sync with SGLang. Would love contribute here as well.

**Comment by SohamRajpure** [comment]
It would also be good to look into making abort_request deterministic so a loop is not required in pause(mode=“abort”)

**Comment by bzantium** [comment]
@SohamRajpure contributions very welcome, please go ahead. Sections 2 and 4 are still open. Section 1 has a minimal drain primitive in #24124, but a paused/quiesce flag (with deterministic abort as you suggested) would be a great follow up.

**Comment by SohamRajpure** [comment]
I'll take a stab at making `abort_request` deterministic while waiting for the maintainer-directed questions in Section 2 to be answered.

**Comment by AlanFokCo** [comment]
We run SGLang behind Alibaba Cloud ACK for model A/B evaluation — same GPU switches between Qwen3-32B and Qwen2.5-32B between evaluation rounds. We hit all three gaps, plus a fourth one that came up once we solved 1-3.

**On point 1 (drain):** We ended up with the same workaround @SohamRajpure described — `pause_generation(mode="abort")` + polling. The non-deterministic abort is the main pain point; we loop with a 100ms sleep until `running_batch` is empty, which is fragile and couples us to scheduler internals. #24124 looks like the right shape. One nuance: after drain, we also need to flush the radix tree cache before touching weights, otherwise the first request after resume can hit prefix-cached KV computed under the old weights (#26792 is the same bug in the persistent HiCache layer).

**On point 3 (sleep/wake):** `release_memory_occupation` + `resume_memory_occupation` works for same-model sleep/wake with `--enable-weights-cpu-backup`. But we found a missing primitive: there's no way to resume with a **different model architecture**. `resume_memory_occupation` restores the exact model that was released — if the new checkpoint has a different `num_hidden_layers`, `hidden_size`, or attention type (GQA vs MLA), there's no path to load it without restarting the process.

In our setup, we implemented a `reload_model()` that does a full in-process model rebuild:

1. Drain + flush cache (your point 1)
2. Destroy CUDA graphs + clear KV pool buffers (`token_to_kv_pool.{k,v}_buffer`)
3. Rebuild `ModelConfig` from new model path
4. `model_runner.load_model()` with the new config
5. `init_memory_pool()` + `init_attention_backend()` (KV dimensions may change)
6. `init_decode_cuda_graph()` with reduced batch sizes for speed
7. Swap tokenizer in `TokenizerManager`

This gets us 2-6s for a 32B cross-architecture switch on 8×A10, vs 60-100s for a process restart. Same-architecture switches preserve CUDA graphs entirely (step 6 skipped → ~0.3s).

The problem is that this currently requires touching ~20 private attributes across `ModelRunner`, `Scheduler`, and `TpWorker`. Would there be interest in an RFC for a supported `reload_model()` API on `ModelRunner`? It would sit alongside the existing `update_weights_from_disk` (same-arch in-place) as the cross-architecture counterpart. Happy to draft the RFC and contribute the implementation.
